src/attribution/onINP.ts

```diff
@@ -25,38 +28,224 @@ import {
+    }
+
+    // Store the grouped render time for this entry for reference later.
+    if (entry.interactionId || entry.entryType === 'first-input') {
+      entryToRenderTimeMap.set(entry, renderTime);
+    }
+  });
+
+  // Queue cleanup of entries that are not part of any INP candidates.
+  if (idleCallback < 0) {
+    idleCallback = requestIdleCallback(cleanupEntries);
+  }
+};
+
+const cleanupEntries = () => {
+  // The list of previous render times is used to handle cases where
+  // events are dispatched out of order. When this happens they're generally
+  // only off by a frame or two, so keeping the most recent 10 should be
+  // more than sufficient.
+  previousRenderTimes = previousRenderTimes.slice(-10);
+
+  // Keep all render times that are part of a pending INP candidate or
+  // that occurred within 2 seconds of the most recent render time.
+  const renderTimesToKeep = new Set(
+    (previousRenderTimes as (number | undefined)[]).concat(
+      longestInteractionList.map((i) => entryToRenderTimeMap.get(i.entries[0])),
+    ),
+  );
+
+  pendingEntriesGroupMap.forEach((_, key) => {
+    renderTimesToKeep.has(key) || pendingEntriesGroupMap.delete(key);
+  });
+
+  // Remove all pending LoAF entries that don't intersect with entries in
+  // the newly cleaned up `pendingEntriesGroupMap`.
+  const loafsToKeep: Set<PerformanceLongAnimationFrameTiming> = new Set();
+  pendingEntriesGroupMap.forEach((group) => {
+    getIntersectingLoAFs(group.startTime, group.processingEnd).forEach(
+      (loaf) => {
+        loafsToKeep.add(loaf);
+      },
+    );
+  });
+  pendingLoAFs = Array.from(loafsToKeep);
+
+  // Reset the idle callback so it can be queued again.
+  idleCallback = -1;
+};
+
+const getIntersectingLoAFs = (
+  start: DOMHighResTimeStamp,
+  end: DOMHighResTimeStamp,
+) => {
+  const intersectingLoAFs = [];
+
+  for (let i = 0, loaf; (loaf = pendingLoAFs[i]); i++) {
+    // If the LoAF ends before the given start time, ignore it.
+    if (loaf.startTime + loaf.duration < start) continue;
+
+    // If the LoAF starts after the given end time, ignore it and all
+    // subsequent pending LoAFs (because they're in time order).
+    if (loaf.startTime > end) break;
+
+    // Still here? If so this LoAF intersects with the interaction.
+    intersectingLoAFs.push(loaf);
   }
-  // Set an empty object if no other attribution has been set.
-  (metric as INPMetricWithAttribution).attribution = {};
+  return intersectingLoAFs;
+};
+
+const attributeINP = (metric: INPMetric): void => {
+  const firstEntry = metric.entries[0];
+  const renderTime = entryToRenderTimeMap.get(firstEntry)!;
+  const group = pendingEntriesGroupMap.get(renderTime)!;
+
+  const processedEventEntries = group.entries;
+  const processingStart = firstEntry.processingStart;
+  const processingEnd = group.processingEnd;
+
+  const longAnimationFrameEntries: PerformanceLongAnimationFrameTiming[] =
+    getIntersectingLoAFs(firstEntry.startTime, processingEnd);
```

Comment:
@mmocny FYI the logic to select which LoAF entries get attributed to INP is here.